src/Microsoft.AspNet.Mvc.Core/Infrastructure/HttpRequestStreamReader.cs

```diff
@@ -0,0 +1,434 @@
+        }
+
+        public override int Read(char[] buffer, int index, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+
+            if (count < 0 || index + count >= buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (_stream == null)
+            {
+                throw new ObjectDisposedException("stream");
+            }
+
+            var charsRead = 0;
+            while (count > 0)
+            {
+                var charsRemaining = _charsRead - _charBufferIndex;
+                if (charsRemaining == 0)
+                {
+                    charsRemaining = ReadIntoBuffer();
+                }
+
+                if (charsRemaining == 0)
+                {
+                    break;  // We're at EOF
+                }
+
+                if (charsRemaining > count)
+                {
+                    charsRemaining = count;
+                }
+
+                Buffer.BlockCopy(
+                    _charBuffer.Array,
+                    (_charBuffer.Offset + _charBufferIndex) * 2,
+                    buffer,
+                    (index + charsRead) * 2,
+                    charsRemaining * 2);
+                _charBufferIndex += charsRemaining;
+
+                charsRead += charsRemaining;
+                count -= charsRemaining;
+
+                // If we got back fewer chars than we asked for, then it's likely the underlying stream is blocked.
+                // Send the data back to the caller so they can process it.
+                if (_isBlocked)
+                {
+                    break;
+                }
+            }
+
+            return charsRead;
+        }
+
+        public override async Task<int> ReadAsync(char[] buffer, int index, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+
+            if (count < 0 || index + count >= buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (_stream == null)
+            {
+                throw new ObjectDisposedException("stream");
+            }
+
+            if (_charBufferIndex == _charsRead && (await ReadIntoBufferAsync().ConfigureAwait(false)) == 0)
+            {
+                return 0;
+            }
+
+            var charsRead = 0;
+            while (count > 0)
+            {
+                // n is the characters available in _charBuffer
+                int n = _charsRead - _charBufferIndex;
```

Comment:
Is there an alternative name for `n` work that works better? `remainingBuffer` or some variation along those lines?